server/src/main/java/io/druid/server/QueryResource.java

```diff
@@ -220,8 +222,17 @@ public void write(OutputStream outputStream) throws IOException, WebApplicationE
                 },
                 contentType
             )
-            .header("X-Druid-Query-Id", queryId)
-            .header("X-Druid-Response-Context", jsonMapper.writeValueAsString(responseContext))
+            .header("X-Druid-Query-Id", queryId);
+
+        //Limit the response-context header, see https://github.com/druid-io/druid/issues/2331
+        String responseCtxString = jsonMapper.writeValueAsString(responseContext);
+        if (responseCtxString.length() > RESPONSE_CTX_HEADER_LEN_LIMIT) {
+          log.warn("Response Context truncated for id [%s] . Full context is [%s].", queryId, responseCtxString);
```

Comment:
hmm... the reason we are logging so that user can see what was full content of the context as some of what is being truncated might be useful meta information.
what is the worst that will happen when we do not truncate the log? maybe some slowness from writing big log line.

i'm not sure if we really need a limit here. what do others think, @xvrl? what would be suggested limit of that?